flask_security/core.py

```diff
@@ -423,28 +440,31 @@ def init_app(self, app, datastore=None, register_blueprint=True,
 
         identity_loaded.connect_via(app)(_on_identity_loaded)
 
-        state = _get_state(app, datastore,
-                           login_form=login_form,
-                           confirm_register_form=confirm_register_form,
-                           register_form=register_form,
-                           forgot_password_form=forgot_password_form,
-                           reset_password_form=reset_password_form,
-                           change_password_form=change_password_form,
-                           send_confirmation_form=send_confirmation_form,
-                           passwordless_login_form=passwordless_login_form,
-                           anonymous_user=anonymous_user)
+        self._state = create_state(app, self.datastore,
+                                   login_form=login_form,
+                                   confirm_register_form=confirm_register_form,
+                                   register_form=register_form,
+                                   forgot_password_form=forgot_password_form,
+                                   reset_password_form=reset_password_form,
+                                   change_password_form=change_password_form,
+                                   send_confirmation_form=send_confirmation_form,
+                                   passwordless_login_form=passwordless_login_form,
+                                   anonymous_user=anonymous_user)
 
         if register_blueprint:
-            app.register_blueprint(create_blueprint(state, __name__))
+            app.register_blueprint(create_blueprint(self._state, __name__))
             app.context_processor(_context_processor)
 
-        state.render_template = self.render_template
-        app.extensions['security'] = state
-
-        return state
+        self._state.render_template = self.render_template
+        app.extensions['security'] = self._state
 
     def render_template(self, *args, **kwargs):
         return render_template(*args, **kwargs)
 
     def __getattr__(self, name):
-        return getattr(self._state, name, None)
+        if self._state is not None:  # shortcut
+            return getattr(self._state, name)
+        ctx = _ctx_stack.top
+        if ctx is not None:
+            self._state = get_state(ctx.app)
+            return getattr(self._state, name)
```

Comment:
Will code below ```ctx = _ctx_stack.top``` ever run? Since ```self._state``` is alway being set in ```__init__``` or ```init_app```, maybe these are additional?